File: src/java/clearvolume/renderer/scenegraph/VolumeNode.java
package clearvolume.renderer.scenegraph;

import cleargl.GLProgram;
import clearvolume.renderer.scenegraph.opencl.OpenCLVolumeRenderer;

import com.jogamp.opengl.GL;
import com.jogamp.opengl.GL2ES3;
import org.apache.commons.io.IOUtils;
import scenery.GeometricalObject;
import java.nio.FloatBuffer;
import java.io.InputStream;
import java.io.IOException;

public class VolumeNode extends GeometricalObject {

    OpenCLVolumeRenderer vrenderer;
    float[] vertices;
    float[] texcoords;

    protected int layer = 0;
    protected String nodeType = "VolumeNode";

    public VolumeNode(OpenCLVolumeRenderer VolumeRenderer) {
        super(3, GL.GL_TRIANGLES);

        vrenderer = VolumeRenderer;
        setRenderer(VolumeRenderer);

        vertices = new float[]{
        -1.0f,-1.0f,0.0f,
        1.0f,-1.0f,0.0f,
        1.0f,1.0f,0.0f,
        -1.0f,-1.0f,0.0f,
        1.0f,1.0f,0.0f,
        -1.0f,1.0f,0.0f
        };

        texcoords = new float[]{0.0f,0.0f,
        1.0f,0.0f,
        1.0f,1.0f,
        0.0f,0.0f,
        1.0f,1.0f,
        0.0f,1.0f
        };
    }

    public boolean init() {
        InputStream lVertexShaderResourceAsStream = VolumeNode.class.getResourceAsStream("shaders/VolumeRender.vs");
        InputStream lFragmentShaderResourceAsStream = VolumeNode.class.getResourceAsStream("shaders/VolumeRender.fs");

        String lVertexShaderSource = "";
        String lFragmentShaderSource = "";

        try {
            lVertexShaderSource = IOUtils.toString(lVertexShaderResourceAsStream, "UTF-8");
            lFragmentShaderSource = IOUtils.toString(lFragmentShaderResourceAsStream, "UTF-8");


            for (int i = 0; i < 1; i++) {
                String lStringToInsert1 = "uniform sampler2D texUnit" + i + "; \n//insertpoint1";
                String lStringToInsert2 = "tempOutColor = max(tempOutColor,texture(texUnit" + i + ", VertexIn.TexCoord));\n//insertpoint2";

                lFragmentShaderSource = lFragmentShaderSource.replace("//insertpoint1",
                        lStringToInsert1);
                //lFragmentShaderSource = lFragmentShaderSource.replace("//insertpoint2",
//                    lStringToInsert2)
            }

            System.out.println(lFragmentShaderSource);

            setProgram(GLProgram.buildProgram(getGL(),
                    lVertexShaderSource,
                    lFragmentShaderSource));
        } catch (IOException e) {
            System.err.println("Could not read shaders");
        }

        super.init();

        setVerticesAndCreateBuffer(FloatBuffer.wrap(vertices));
        setTextureCoordsAndCreateBuffer(FloatBuffer.wrap(texcoords));

        return true;
    }

    public int getLayer() {
        return layer;
    }

    public void draw() {
        getGL().glDisable(GL.GL_CULL_FACE);
        getGL().glEnable(GL.GL_BLEND);
        getGL().glBlendFunc(GL.GL_ONE, GL.GL_ONE);
        getGL().glBlendEquation(GL2ES3.GL_MAX);
        getGL().glDisable(GL.GL_DEPTH_TEST);

        vrenderer.renderVolume(
                    getModelView().clone().transpose().getFloatArray(),
                    getModelView().clone().transpose().getFloatArray(), layer);

        vrenderer.clearChangeOfVolumeParametersFlag();
        getProgram().getUniform("texUnit0").setInt(layer);
        getProgram().getUniform("layer").setInt(layer);

        super.draw();
    }
}

File: src/java/clearvolume/renderer/cleargl/VolumeNode.kt
package clearvolume.renderer.cleargl;

import cleargl.GLProgram
import cleargl.scenegraph.GeometricalObject
import clearvolume.renderer.opencl.OpenCLVolumeRenderer
import com.jogamp.opengl.GL
import com.jogamp.opengl.GL2ES3
import org.apache.commons.io.IOUtils
import java.nio.FloatBuffer

class VolumeNode(VolumeRenderer: OpenCLVolumeRenderer) : GeometricalObject(3, GL.GL_TRIANGLES) {

    var vrenderer: OpenCLVolumeRenderer
    var vertices: FloatArray
    var texcoords: FloatArray

    var layer: Int = 0

    override var nodeType: String = "VolumeNode"

    init {
        vrenderer = VolumeRenderer
        renderer = VolumeRenderer

        vertices = floatArrayOf(
                -1.0f, -1.0f, 0.0f,
                1.0f, -1.0f, 0.0f,
                1.0f, 1.0f, 0.0f,
                -1.0f, -1.0f, 0.0f,
                1.0f, 1.0f, 0.0f,
                -1.0f, 1.0f, 0.0f
        )

        texcoords = floatArrayOf(0.0f, 0.0f,
                1.0f, 0.0f,
                1.0f, 1.0f,
                0.0f, 0.0f,
                1.0f, 1.0f,
                0.0f, 1.0f
        )
    }

    override fun init(): Boolean {
        val lVertexShaderResourceAsStream = VolumeNode::class.java.getResourceAsStream("shaders/VolumeRender.vs")
        val lFragmentShaderResourceAsStream = VolumeNode::class.java.getResourceAsStream("shaders/VolumeRender.fs")

        val lVertexShaderSource = IOUtils.toString(lVertexShaderResourceAsStream, "UTF-8")
        var lFragmentShaderSource = IOUtils.toString(lFragmentShaderResourceAsStream, "UTF-8")

        for (i in 1..1) {
            val lStringToInsert1 = "uniform sampler2D texUnit%d; \n//insertpoint1".format(i)
            val lStringToInsert2 = "tempOutColor = max(tempOutColor,texture(texUnit%d, VertexIn.TexCoord));\n//insertpoint2".format(i)

            lFragmentShaderSource = lFragmentShaderSource.replace("//insertpoint1",
                    lStringToInsert1)
            //lFragmentShaderSource = lFragmentShaderSource.replace("//insertpoint2",
//                    lStringToInsert2)
        }

        System.out.println(lFragmentShaderSource)

        program = GLProgram.buildProgram(gl!!,
                lVertexShaderSource,
                lFragmentShaderSource)

        super.init()

        setVerticesAndCreateBuffer(FloatBuffer.wrap(vertices))
        setTextureCoordsAndCreateBuffer(FloatBuffer.wrap(texcoords))

        return true;
    }

    override fun draw() {
        gl!!.glDisable(GL.GL_CULL_FACE)
        gl!!.glEnable(GL.GL_BLEND)
        gl!!.glBlendFunc(GL.GL_ONE, GL.GL_ONE)
        gl!!.glBlendEquation(GL2ES3.GL_MAX)
        gl!!.glDisable(GL.GL_DEPTH_TEST)

        vrenderer.renderVolume(
                    modelView!!.clone().transpose().floatArray,
                    projection!!.clone().transpose().floatArray, layer)

        vrenderer.clearChangeOfVolumeParametersFlag()
        program!!.getUniform("texUnit0").setInt(layer)
        program!!.getUniform("layer").setInt(layer)

        super.draw()
    }
}